In [1]:
from utils.metrics import MetricsComputation

## Fields
 - c_digo_dane_departamento= Código Dane departamento
 - departamento=Departamento
 - c_digo_dane_municipio=Código Dane municipio
  - municipio=Municipio
  - grupo_cultivo=Grupo cultivo
  - subgrupo=Subgrupo
  - cultivo=Cultivo
  - desagregaci_n_cultivo=Desagregación cultivo
 - a_o=Año
  - periodo=Periodo
  - rea_sembrada=Área sembrada
  - rea_cosechada=Área cosechada
  - producci_n=Producción (toneladas)
  - rendimiento=Rendimiento
  - ciclo_del_cultivo=Ciclo del cultivo
  - estado_f_sico_del_cultivo=Estado físico del cultivo
  - c_digo_del_cultivo=Código del cultivo
  - nombre_cient_fico_del_cultivo=Nombre científico del cultivo

In [2]:
file_path='data/transformed/agriculture.csv'

In [3]:
eda = MetricsComputation(file_path)
eda.display_head()

First 5 records:


c_d_dep        departamento  c_d_mun   municipio grupo_de_cultivo  \
0       25        Cundinamarca    25754      Soacha       Hortalizas   
1       25        Cundinamarca    25214        Cota       Hortalizas   
2       54  Norte de santander    54405  Los patios       Hortalizas   
3       54  Norte de santander    54518    Pamplona       Hortalizas   
4       68           Santander    68377  La belleza       Hortalizas   

  subgrupo_de_cultivo cultivo desagregaci_n_regional_y   año periodo  \
0              Acelga  Acelga                   Acelga  2006   2006b   
1              Acelga  Acelga                   Acelga  2006   2006b   
2              Acelga  Acelga                   Acelga  2006   2006b   
3              Acelga  Acelga                   Acelga  2006   2006b   
4              Acelga  Acelga                   Acelga  2006   2006b   

   rea_sembrada_ha  rea_cosechada_ha  producci_n_t  rendimiento_t_ha  \
0             82.0              80.0        1440.0             18.00   
1              2.0               2.0          26.0             17.33   
2              3.0               3.0          48.0             16.00   
3              1.0               1.0           5.0             10.00   
4              1.0               1.0           6.0              6.00   

  estado_fisico_produccion nombre_cientifico ciclo_de_cultivo  \
0             Fruto fresco     Beta vulgaris      Transitorio   
1             Fruto fresco     Beta vulgaris      Transitorio   
2             Fruto fresco     Beta vulgaris      Transitorio   
3             Fruto fresco     Beta vulgaris      Transitorio   
4             Fruto fresco     Beta vulgaris      Transitorio   

   c_digo_del_cultivo  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN

In [5]:
eda.display_info()

Dataset information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 291770 entries, 0 to 291769
Data columns (total 18 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   c_d_dep                   291770 non-null  int64  
 1   departamento              291770 non-null  object 
 2   c_d_mun                   291770 non-null  int64  
 3   municipio                 291769 non-null  object 
 4   grupo_de_cultivo          291770 non-null  object 
 5   subgrupo_de_cultivo       291770 non-null  object 
 6   cultivo                   291770 non-null  object 
 7   desagregaci_n_regional_y  291770 non-null  object 
 8   año                       291770 non-null  int64  
 9   periodo                   291770 non-null  object 
 10  rea_sembrada_ha           291770 non-null  float64
 11  rea_cosechada_ha          291770 non-null  float64
 12  producci_n_t              291770 non-null  float64
 13  rendimiento_t_ha       

None

In [6]:
eda.missing_values()

Percentage missing values per column (%):


c_d_dep                       0.0%
departamento                  0.0%
c_d_mun                       0.0%
municipio                     0.0%
grupo_de_cultivo              0.0%
subgrupo_de_cultivo           0.0%
cultivo                       0.0%
desagregaci_n_regional_y      0.0%
año                           0.0%
periodo                       0.0%
rea_sembrada_ha               0.0%
rea_cosechada_ha              0.0%
producci_n_t                  0.0%
rendimiento_t_ha             1.17%
estado_fisico_produccion      0.0%
nombre_cientifico            0.98%
ciclo_de_cultivo              0.0%
c_digo_del_cultivo          68.55%
dtype: object

In [7]:
eda.zero_values()

Percentage zero values per column (%):


c_d_dep                      0.0%
departamento                 0.0%
c_d_mun                      0.0%
municipio                    0.0%
grupo_de_cultivo             0.0%
subgrupo_de_cultivo          0.0%
cultivo                      0.0%
desagregaci_n_regional_y     0.0%
año                          0.0%
periodo                      0.0%
rea_sembrada_ha             0.91%
rea_cosechada_ha            2.61%
producci_n_t                2.43%
rendimiento_t_ha            1.13%
estado_fisico_produccion     0.0%
nombre_cientifico            0.0%
ciclo_de_cultivo             0.0%
c_digo_del_cultivo           0.0%
dtype: object

- La mayoría de las columnas tienen una cobertura completa de datos (291,770 entradas) con muy pocos valores faltantes.
- Las columnas rendimiento_t_ha (1.17% de valores faltantes), nombre_cientifico (0.98% de valores faltantes) y c_digo_del_cultivo (68.55% de valores faltantes) presentan valores ausentes, siendo la última la que tiene un porcentaje significativamente alto. La columna c_digo_del_cultivo presenta una gran cantidad de valores faltantes (68.55%). Esto puede impactar negativamente en los análisis que requieren esta información específica para clasificar los cultivos por lo cual se debe indagar con el negocio el motivo de dichos valores ausentes.
- Solo las columnas rea_sembrada_ha, rea_cosechada_ha, producci_n_t y rendimiento_t_ha tienen un porcentaje de valores cero, con 0.91%, 2.61%, 2.43% y 1.13% respectivamente.


A pesar de la presencia de algunos valores cero y faltantes, la mayoría de los datos están completos, lo que sugiere que el conjunto de datos es bastante robusto para análisis detallados.

In [8]:
eda.duplicate_rows()

Total of duplicated rows:


Empty DataFrame
Columns: [c_d_dep, departamento, c_d_mun, municipio, grupo_de_cultivo, subgrupo_de_cultivo, cultivo, desagregaci_n_regional_y, año, periodo, rea_sembrada_ha, rea_cosechada_ha, producci_n_t, rendimiento_t_ha, estado_fisico_produccion, nombre_cientifico, ciclo_de_cultivo, c_digo_del_cultivo]
Index: []

- No existen registros duplicados en la base de datos lo que garantiza que cada entrada es única.

In [9]:
eda.unique_values()

Unique values per column:


{'c_d_dep': 32,
 'departamento': 32,
 'c_d_mun': 1107,
 'municipio': 1089,
 'grupo_de_cultivo': 16,
 'subgrupo_de_cultivo': 139,
 'cultivo': 246,
 'desagregaci_n_regional_y': 317,
 'año': 17,
 'periodo': 48,
 'rea_sembrada_ha': 10418,
 'rea_cosechada_ha': 9847,
 'producci_n_t': 22910,
 'rendimiento_t_ha': 4065,
 'estado_fisico_produccion': 31,
 'nombre_cientifico': 277,
 'ciclo_de_cultivo': 3,
 'c_digo_del_cultivo': 190}

- Los datos abarcan 32 departamentos y 1,107 municipios, lo que sugiere una amplia cobertura geográfica dentro de la región estudiada.
- Se identificaron 16 grupos de cultivo, con un total de 139 subgrupos y 246 cultivos específicos. Esto indica una diversidad agrícola significativa en la zona.
- Los datos abarcan un período de 17 años, lo que permite un análisis detallado a lo largo del tiempo.
- Hay 31 estados físicos de producción diferentes, lo que sugiere una variedad de condiciones en las cuales se registran los datos de producción.

En general, estos hallazgos destacan la diversidad y complejidad de la actividad agrícola en la región estudiada, lo que proporciona una base sólida para análisis y toma de decisiones en el sector agrícola.

In [10]:
eda.describe_numeric()

Descriptive statistics of numeric variables:


c_d_dep        c_d_mun            año  rea_sembrada_ha  \
count  291770.000000  291770.000000  291770.000000    291770.000000   
mean       40.045604   40470.473239    2014.959992       275.175895   
std        25.207168   25185.014235       4.806329      1166.046467   
min         5.000000    5001.000000    2006.000000         0.000000   
25%        17.000000   17446.000000    2011.000000         8.000000   
50%        41.000000   41306.000000    2015.000000        30.000000   
75%        66.000000   66594.000000    2019.000000       135.000000   
max        99.000000   99773.000000    2022.000000     47403.000000   

       rea_cosechada_ha  producci_n_t  rendimiento_t_ha  c_digo_del_cultivo  
count     291770.000000  2.917700e+05     288350.000000        9.177000e+04  
mean         239.962379  2.929600e+03          9.577996        1.436073e+06  
std         1026.571249  4.665096e+04         15.364008        4.850887e+05  
min            0.000000  0.000000e+00          0.000000        1.010101e+06  
25%            7.000000  2.800000e+01          1.500000        1.051700e+06  
50%           25.000000  1.250000e+02          5.000000        1.080400e+06  
75%          113.000000  6.000000e+02         12.000000        2.041000e+06  
max        46264.000000  4.776340e+06        640.000000        2.080300e+06

- La desviacion estandar para rea_sembrada_ha y rea_cosechada_ha está considerablemente alejada de la media respectivamente, lo que sugiere la presencia de numerosos valores atípicos en los datos registrados en estos campos. 

In [12]:
group_by_columns=['año', 'departamento', 'grupo_de_cultivo']
aggregation_columns = {'count': ('grupo_de_cultivo', 'count'), 'avg_production': ('producci_n_t', 'mean')}
sort_by_column='avg_production'
grouped_table=eda.grouped_table(group_by_columns,aggregation_columns,sort_by_column)
grouped_table

año     departamento                   grupo_de_cultivo  count  \
0  2021  Valle del cauca  Cultivos tropicales tradicionales    134   
1  2019  Valle del cauca  Cultivos tropicales tradicionales    134   
2  2022  Valle del cauca  Cultivos tropicales tradicionales    138   
3  2020  Valle del cauca  Cultivos tropicales tradicionales    145   
4  2018  Valle del cauca                  Otros permanentes    136   
5  2014  Valle del cauca                  Otros permanentes    132   
6  2015  Valle del cauca                  Otros permanentes    132   
7  2017  Valle del cauca                  Otros permanentes    136   
8  2016  Valle del cauca                  Otros permanentes    132   
9  2009  Valle del cauca                  Otros permanentes    138   

   avg_production  
0   185209.543582  
1   176463.505672  
2   175471.804275  
3   163673.390138  
4   156487.139706  
5   149239.901515  
6   148375.977273  
7   143544.007353  
8   141073.704545  
9   136658.478261

- Los 'cultivos tropicales tradicionales' en el departamento del Valle del Cauca fueron los que mayor producción promedio tuvieron por año (2021).

In [14]:
group_by_columns=['grupo_de_cultivo']
aggregation_columns = {'count': ('grupo_de_cultivo', 'count'), 'sum_production': ('producci_n_t', 'sum')}
sort_by_column='sum_production'
grouped_table=eda.grouped_table(group_by_columns,aggregation_columns,sort_by_column)
grouped_table

grupo_de_cultivo  count  sum_production
0                  Otros permanentes  21813    3.049079e+08
1  Cultivos tropicales tradicionales   8292    1.747696e+08
2                           Frutales  76730    1.161920e+08
3              Tuberculos y platanos  26564    9.864360e+07
4                           Cereales  51394    7.127700e+07
5                         Hortalizas  47665    2.990462e+07
6                Raices y tuberculos  11831    2.735819e+07
7                        Oleaginosas   3128    2.423346e+07
8                        Leguminosas  36481    5.426690e+06
9                             Fibras   1977    1.104701e+06

- Los grupos de cultivos tropicales tradicionales, frutales y raíces/tubérculos han registrado las mayores producciones en la historia, con 44,992,148 toneladas, 12,886,524 toneladas y 6,632,276 toneladas respectivamente.

In [16]:
df = eda.data
df_filtered = df[df['año']==2022]
eda.update_dataframe(df_filtered)

group_by_columns=['año', 'grupo_de_cultivo']
aggregation_columns = {'count': ('grupo_de_cultivo', 'count'), 'avg_production': ('producci_n_t', 'mean')}
sort_by_column='avg_production'
grouped_table=eda.grouped_table(group_by_columns,aggregation_columns,sort_by_column)
grouped_table

año                                   grupo_de_cultivo  count  \
0  2022                  Cultivos tropicales tradicionales   2086   
1  2022                                        Oleaginosas    288   
2  2022                                Raices y tuberculos   3153   
3  2022                                           Frutales   7304   
4  2022                                           Cereales   4614   
5  2022                                         Hortalizas   4540   
6  2022                                        Leguminosas   2601   
7  2022  Cultivos para condimentos, bebidas medicinales...    216   

   avg_production  
0    21568.623600  
1     8624.215903  
2     2103.481383  
3     1764.310607  
4     1119.459365  
5      606.247055  
6      107.447970  
7      102.826620

- La producción promedio más alta se observa en los cultivos tropicales tradicionales con un promedio de 21,568.62 toneladas, seguidos de las oleaginosas con un promedio de 8,624.22 toneladas y las raíces y tubérculos con un promedio de 2,103.48 toneladas. Estos cultivos muestran una producción significativamente más alta en comparación con otros tipos de cultivos en el año 2022.

In [13]:
group_by_columns=['departamento']
aggregation_columns = { 'avg_production': ('producci_n_t', 'mean')}
sort_by_column='avg_production'
grouped_table=eda.grouped_table(group_by_columns,aggregation_columns,sort_by_column)
grouped_table

departamento  avg_production
0  Valle del cauca    16179.284447
1           Arauca     9133.944309
2            Cauca     6202.676315
3             Meta     5242.790015
4         Casanare     4326.628187
5        Risaralda     2821.982518
6           Caldas     2793.566067
7         Guaviare     2598.727083
8           Tolima     2429.743456
9        Magdalena     2263.001061

In [14]:
df=eda.data
avg_production = df.groupby('departamento').agg({'producci_n_t':'mean'}).sort_values(by= 'producci_n_t',ascending=False).reset_index()
avg_production

departamento  producci_n_t
0             Valle del cauca  16179.284447
1                      Arauca   9133.944309
2                       Cauca   6202.676315
3                        Meta   5242.790015
4                    Casanare   4326.628187
5                   Risaralda   2821.982518
6                      Caldas   2793.566067
7                    Guaviare   2598.727083
8                      Tolima   2429.743456
9                   Magdalena   2263.001061
10                    Quindio   2250.263967
11                  Santander   2223.639496
12                  Antioquia   2189.239816
13                    Cordoba   2072.702274
14                    Caqueta   2016.225860
15               Cundinamarca   1852.291289
16                      Cesar   1740.782432
17                     Nariño   1697.790634
18                    Vichada   1654.751563
19                    Bolivar   1560.987284
20                     Boyaca   1448.807960
21         Norte de santander   1306.236968
22                      Choco   1253.354429
23                      Huila   1217.178551
24                      Sucre   1053.722459
25                   Putumayo    961.004028
26                 La guajira    595.566582
27                    Guainia    530.563333
28                  Atlantico    339.173300
29                   Amazonas    292.688028
30                     Vaupes    279.354286
31  San andres y providencias     24.588333